src/aproxy.py

```diff
@@ -197,27 +200,45 @@ def _convert_ports_to_ranges(cls, ports: List[str]) -> List[tuple]:
         """Convert all port values to sorted ranges."""
         ranges: List[tuple] = []
         for item in ports:
-            if "-" in item:
-                try:
-                    start, end = map(int, item.split("-", 1))
-                except ValueError as exc:
-                    raise ValueError(f"invalid port range: {item}") from exc
-                if not 1 <= start <= end <= 65535:
-                    raise ValueError(f"port range must be between 1 and 65535 instead of {item}")
-                ranges.append((start, end))
-            else:
-                try:
-                    port = int(item)
-                except ValueError as exc:
-                    raise ValueError(
-                        f"port value must be of type integer instead of {item}"
-                    ) from exc
-                if not 1 <= port <= 65535:
-                    raise ValueError(f"port must be between 1 and 65535 instead of {item}")
-                ranges.append((port, port))
+            if "-" not in item:
+                item = f"{item}-{item}"
+            try:
+                start, end = map(int, item.split("-", 1))
+            except ValueError as exc:
+                raise ValueError(f"invalid port range: {item}") from exc
+            if not 1 <= start <= end <= 65535:
+                raise ValueError(f"port range must be between 1 and 65535 instead of {item}")
+            ranges.append((start, end))
 
         return ranges
 
+    @classmethod
+    def _get_principal_proxy_address(cls, relation: ops.model.Relation | None) -> str:
+        """Get the principal charm's proxy address from the relation data.
+
+        Args:
+            relation: The Juju relation to the principal charm.
+
+        Returns:
+            The proxy address (hostname or IP) if set, else an empty string.
+        """
+        proxy_conf = ""
+        if not relation:
+            return proxy_conf
+
+        app_data = relation.data.get(relation.app)
+        unit_data = next(iter(relation.units), None)
+        if app_data and ("juju-https-proxy" in app_data or "juju-http-proxy" in app_data):
+            proxy_conf = app_data.get("juju-https-proxy") or app_data.get("juju-http-proxy") or ""
+        elif unit_data:
+            unit_conf = relation.data[unit_data]
+            proxy_conf = (
+                unit_conf.get("juju-https-proxy") or unit_conf.get("juju-http-proxy") or ""
+            )
```

Comment:
I think there's been a misunderstanding here. We can get the proxy settings from the model configuration which are injected as environment variables in your charm.
See:
https://github.com/canonical/synapse-operator/blob/2/main/src/state/charm_state.py#L269-L282